cluster_generator/data_structures.py

```diff
@@ -1,100 +1,364 @@
+        )
+        _relative_bbox = self.bbox - center.d.reshape((3, 1))
+
+        mylog.info(f"Adding {model} to {self.__str__()}")
+        mylog.info(
+            f"\tPos: {[np.round(j,decimals=2) for j in center.d]} kpc, Vel: {[np.round(j,decimals=2) for j in velocity.to_value('km/s')]} km/s"
+        )
+
+        # setup the coordinates to maximize ease of interpolation.
+        _rr = model["radius"].to_value("kpc")
+        _dens = model["density"].to_value(fields["density"])
+
+        with self.open() as ythdf5_io:
+            for field, unit in tqdm(
+                fields.items(),
+                desc=f"Writing {model} to YTHDF5",
+                leave=False,
+                disable=(not cgparams["system"]["display"]["progress_bars"]),
+            ):
+                if "momentum_density" in field:
+                    # manage the momentum density.
+                    _momentum_index = {"x": 0, "y": 1, "z": 2}[field[-1]]
+                    _yy = _dens * velocity[_momentum_index].to_value(unit)
+                elif field in model.fields:
+                    _yy = model[field].to_value(unit)
+                else:
+                    mylog.warning(
+                        f"Failed to write model data for {field}; the field doesn't exist in {model}."
+                    )
+                    continue
+
+                self._add_field(
+                    ythdf5_io, _rr, _yy, field, _relative_bbox, self.chunkmap
+                )
+
+            mylog.info(f"Core fields of {model} where written to {self}.")
+
+            ythdf5_io.attrs["model_count"] += 1
+
+    def add_ICs(self, ics):
+        """
+        Add an entire :py:class:`ics.ClusterICs` instance to the :py:class:`data_structures.YTHDF5` buffer.
+
+        Parameters
+        ----------
+        ics: :py:class:`ics.ClusterICs`
+            The initial conditions to add to the HDF5 buffer.
+        """
+        mylog.info(f"Adding {ics.basename} to {self}.")
+
+        for ic_id, ic_model in enumerate(
+            tqdm(
+                ics.profiles,
+                desc=f"Writing {ics} to YTHDF5",
+                position=0,
+                leave=False,
+                disable=(not cgparams["system"]["display"]["progress_bars"]),
+            )
+        ):
+            model = ClusterModel.from_h5_file(ic_model)
+            center, velocity = ics.center[ic_id], ics.velocity[ic_id]
+
+            self.add_model(model, center, velocity)
+
+    def _add_field(self, fileio, r, y, fieldname, bbox, chunkmap):
+        # Interpolate and pass to the Cython level.
+        _, _, _, _, _, k, _, n, t, c, _, _, _, _ = dfitpack.fpcurf0(
+            r, y, 3, w=None, xb=r[0], xe=r[-1], s=0.0
+        )
+        _buffer_obj = fileio["grid"][fieldname]
+        dump_field_to_hdf5(
+            _buffer_obj, bbox, self.domain_dimensions, chunkmap, t, c, k, fieldname
+        )
+
+
+if __name__ == "__main__":
+    with YTHDF5.temporary() as temp:
```

Comment:
:laughing: I'm not sure I understand the utility of it anymore. I think when I wrote it I was thinking about keeping the working directory from getting cluttered if you needed many of them, but this comes with other issues.

If the user makes an error inside the ``with`` statement, it could live in their temp space forever which is also not ideal. I think it makes more sense to let people implement this logic for themselves... 

I've removed it. Easy enough to add back if it ever makes sense to do so...